### DISCUSSION TITLE
Windows DLL

### DISCUSSION THREAD

**Discussion by ModeZt** [open-telemetry/opentelemetry-cpp#1921] [Q&A] [answered]
As stated in https://github.com/open-telemetry/opentelemetry-cpp/blob/main/INSTALL.md you are not allowed to build a windows DLL:

> For now, OpenTelemetry C++ targets need to be statically linked into the Windows applications.

Can I make a workaround like this:
1) Create a WrapperDll that is statically linked with OTEL
2) Dinamically load WrapperDll in my application?

I've inspected the topic: https://github.com/open-telemetry/opentelemetry-cpp/discussions/1269 and still have a hope that my workaround could work

**Comment by owent** [comment, marked as answer]
> As stated in https://github.com/open-telemetry/opentelemetry-cpp/blob/main/INSTALL.md you are not allowed to build a windows DLL:
> 
> > For now, OpenTelemetry C++ targets need to be statically linked into the Windows applications.
> 
> Can I make a workaround like this:
> 
>     1. Create a WrapperDll that is statically linked with OTEL
> 
>     2. Dinamically load WrapperDll in my application?
> 
> 
> I've inspected the topic: #1269 and still have a hope that my workaround could work

Yes, if you don't export any API or data struct of otel-cpp in your WrapperDll.

**Comment by malkia** [comment]
Coincidentally I'm in need of a DLL version for Windows too. I did several iterations on my own, and my last attempt is here https://github.com/malkia/opentelemetry-cpp - it builds a single `otel_sdk.dll` and adds various `OPENTELEMETRY_API` where needed.

Also my "hack" is for `bazel` only. I don't have much experience with CMake.

More docs here - https://github.com/malkia/opentelemetry-cpp/blob/main/dll.md 
It's heavy WIP, I'm working on adding packaging into single zip file.

Most likely I probably took some really wrong decision here, but my goals were two:
  - Fix the singletons (OPENTELEMETRY_API_SINGLETON is noop for MSVC)
  - Have .dll in order to avoid needing to link to extra dependencies.

To test it quickly from my branch:
```
bazel test //:with_dll=true ...
```

Note: Jaeger's test/build are ignored, also the `.bazelrc` may have to be changed before use.

**Comment by lalitb** [comment]
@malkia Coincidently there is some work done as part of #1932 if this is something you are looking for?

**Reply by malkia** [reply]
LOL. This is awesome :) - Thanks!

**Reply by malkia** [reply]
I took a brief look to compare with my approach... In my version, nothing in the "api", is marked with OPENTELEMETRY_API (e.g. __declspec(dllimport/export)).. only the "sdk" and "exporter/ext" classes. I also define OPENTELEMETRY_API_SINGLETON to be the same as the API (instead of __declspec(noinline)). I also had to add several deletions to copy consturctors/operators, where unique_ptr were part of the class, but was able to get all testst (except jaeger, which I need to find way to compile). As bonus, singleton_test works, where in the static (MSVC) version it does not.

I'm really looking forward towards a DLL version landing, as this would easy our own adoption.

**Reply by lalitb** [reply]
@malkia Feel free to comment directly on the PR if you see any improvements.
